### PR TITLE
RHEL 9: Filesystem customizations for edge-raw-image

### PR DESCRIPTION
**Pull request by say-paul**
This PR enables File system customization for edge-raw-image.

### CODE REVIEW

**Comment by say-paul** [comment]
having the following error:
```
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="    Failed to open file \"/sys/fs/selinux/checkreqprot\": Read-only file system" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="    LVM2: using vg name '25287c67-6499-43cd-a25f-a80f90f36cb0'" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      Physical volume \"/dev/mapper/osbuild-luks-46d73afa-d0b3-4c68-860c-ca1c52f87211\" successfully created." jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      Creating devices file /etc/lvm/devices/system.devices" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      Volume group \"25287c67-6499-43cd-a25f-a80f90f36cb0\" successfully created" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      WARNING: Logical volume 25287c67-6499-43cd-a25f-a80f90f36cb0/rootlv not zeroed." jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      Logical volume \"rootlv\" created." jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      Volume group \"25287c67-6499-43cd-a25f-a80f90f36cb0\" has insufficient free space (252 extents): 256 required." jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="    Traceback (most recent call last):" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      File \"/run/osbuild/bin/org.osbuild.lvm2.create\", line 110, in <module>" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="        ret = main(args[\"devices\"], args[\"options\"])" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      File \"/run/osbuild/bin/org.osbuild.lvm2.create\", line 105, in main" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="        subprocess.run(cmd, encoding='utf8', check=True)" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="      File \"/usr/lib64/python3.9/subprocess.py\", line 528, in run" jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="        raise CalledProcessError(retcode, process.args," jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
Nov 13 18:03:33 saypaul-thinkpadx1carbongen9.pnq.csb osbuild-worker[1450]: time="2023-11-13T18:03:33+05:30" level=info msg="    subprocess.CalledProcessError: Command '['lvcreate', '-an', '--size', '1073741824B', '-n', 'foo_barlv', '25287c67-6499-43cd-a25f-a80f90f36cb0']' returned non-zero exit status 5." jobId=3230a5f5-ee82-416e-9be3-11e7089cc948
```

with blueprint:
```
name = "rhel93-fs-image"
description = "rhel93-fs-image"
version = "0.0.0"
packages = []
modules = []
groups = []
distro = "rhel-93"

[customizations]

[[customizations.filesystem]]
mountpoint = "/foo/bar"
size = 1
[customizations.ignition]
[customizations.ignition.firstboot]
url = "http://192.168.122.1:8001/config.ign"
```

**Comment by say-paul** [comment]
blueprint for customization:

```
~/composed_images $ cat ~/blueprints/rhel93-multiple-v2-fs-image.toml 
name = "rhel93-multiple-v2-fs-image"
description = "rhel93-fs-multiple-v2-image"
distro = "rhel-93"

[[customizations.filesystem]]
mountpoint = "/foo/bar"
size=2147483648

[[customizations.filesystem]]
mountpoint = "/foo"
size=8589934592


[[customizations.filesystem]]
mountpoint = "/var/myfiles"
size= "1 GiB"

[customizations.ignition.firstboot]
url = "http://192.168.122.1:8001/config.ign"

~/composed_images $ composer-cli blueprints show rhel93-multiple-v2-fs-image 
name = "rhel93-multiple-v2-fs-image"
description = "rhel93-fs-multiple-v2-image"
version = "0.0.0"
packages = []
modules = []
groups = []
distro = "rhel-93"

[customizations]

[[customizations.filesystem]]
mountpoint = "/foo/bar"
size = 2147483648

[[customizations.filesystem]]
mountpoint = "/foo"
size = 8589934592

[[customizations.filesystem]]
mountpoint = "/var/myfiles"
size = 1073741824
[customizations.ignition]
[customizations.ignition.firstboot]
url = "http://192.168.122.1:8001/config.ign"


```

**Comment by say-paul** [comment]
composed image
image

**Comment by achilleas-k** [comment]
I found the issue.  It's a tricky detail with the way the partition table is described internally.  Volume groups are sizeable containers, they need to have a size defined.  But it seems we don't do this in our static partition tables, so the volume group starts off as 0 bytes and is then grown to fill the space of the required total image size.  This worked fine so far, but once the logical volume sizes add up to more than the image size, we run out of space.

The proper solution here is to have logic which resizes the volume group to fit the logical volumes it contains and also have it calculate the size of those LVs when needed.  It would basically be the same logic we have for the partition table as a whole: the vg has a "desired size", if the sum of lv sizes inside it is bigger than that desired size we grow it, if it's not then it's set to that size.

I'll get a fix (and some tests) done asap.

**Comment by say-paul** [comment]
Interesting details, can we pre-calculate the space required  and create the VG accordingly, and then grow what bootlv to fit the reaming space in imgae.disk space is more.

**Comment by achilleas-k** [comment]
> Interesting details, can we pre-calculate the space required and create the VG accordingly, and then grow rootlv to fit the remaining space in image.disk_space.

yeah that's the plan.  I'm working on that today.

File: internal/pathpolicy/path_policy.go
Comment (achilleas-k) [inline internal/pathpolicy/path_policy.go:33]: Combined with the usage below, this will mutate the global default policy (`entry_1`) and basically add a `deny` rule for `/ostree` to the default.

**Review by achilleas-k — CHANGES_REQUESTED** [review]

File: internal/pathpolicy/path_policy.go
Comment (say-paul) [reply internal/pathpolicy/path_policy.go]: Fixed, add if does not exists

File: internal/pathpolicy/path_policy.go
Comment (achilleas-k) [inline internal/pathpolicy/path_policy.go:34]: This doesn't fix the issue.  You're still modifying `entry_1` which is the global `defaultMountpointPolicies` when you're creating the `OstreeMountpointPolicies`.

Run this test, for example, to see the issue:
```go
func TestPathPolicyMerge(t *testing.T) {
	_, ok := defaultMountpointPolicies["/ostree"]
	assert.False(t, ok, "/ostree found in defaultMountpointPolicies")
}
```
You can copy the first map and then merge the second into it to make this work as intended, or make it so that the function explicitly modifies one of the two maps, and use it that way, but that might be trickier to communicate.

Also, an aside I didn't mention earlier: `entry_1` and `entry_2` aren't great variable names here and they're not idiomatic for the language (underscores).

Comment (achilleas-k) [reply internal/pathpolicy/path_policy.go:34]: Your two versions do raise an interesting question though.  Which policy map wins when both being merged have a policy for the same path?  In the first version it was the second entry and now in the second version it's the first.  This should also be made explicit, probably in the docstring of the function, however it ends up working.

**Review by thozza — COMMENTED** [review]
Thanks for your work. I added a few comments inline.

On a general note, I'm not a big fan of merging the policies in the way you implemented it. I admit that this is probably a bit of a personal preference, but I feel like the merging makes the code more complex. In addition, the destination of the merge function is modified in-place and can override parts of the source policy (which is not needed AFAICT and it is also not explicitly documented and not covered by the unit test).

My preference would be to not touch the default path policy and how it is checked. Then in the `checkOptions()` function, you could simply do something like this after the default policy check:

```go
...
err := blueprint.CheckMountpointsPolicy(mountpoints, pathpolicy.MountpointPolicies)
if err != nil {
	return warnings, err
}

if t.rpmOstree {
	err := blueprint.CheckMountpointsPolicy(mountpoints, pathpolicy.OstreeMountpointPolicies)
	if err != nil {
		return warnings, err
	}
}
...
```

This way, the ostree mountpoint policy would be really just an add-on to the default polisy and very few new code and logic would need to be added.

**Comment by say-paul** [comment]
@thozza thank you the suggestions, its much simpler to implement.

File: pkg/distro/rhel9/imagetype.go
Comment (thozza) [inline pkg/distro/rhel9/imagetype.go:390]: I would keep this here and just relax the condition to not apply to `edge-raw-image`. The reason is that doing this check after the default mountpoint policy may result in a situation when the user would get an error related to forbidden mountpoint path for any ostree image type, even one that does not support any FS customizations. That may be confusing and make it look as if the FS customization was supported for the given ostree image type.

Later on, you IMO don't need to check the image type, only if it is ostree-based, because the check for unsupported ostree image types would return from the function before reaching the code.

File: internal/pathpolicy/policies.go
Comment (thozza) [reply internal/pathpolicy/policies.go:52]: Yeah, that should be fine :+1:

File: test/configs/filesystem-customizations.json
Comment (say-paul) [reply test/configs/filesystem-customizations.json]: renamed the file system to `test/configs/ostree-filesystem-customizations.json` and extended the customization too for the comment below.

File: internal/pathpolicy/policies.go
Comment (say-paul) [reply internal/pathpolicy/policies.go]: [updated here](https://github.com/osbuild/images/pull/255#discussion_r1415537208)

File: pkg/distro/rhel9/imagetype.go
Comment (thozza) [reply pkg/distro/rhel9/imagetype.go:388]: I know, but as I mentioned, it is too late, because the code can return with FS customization related error on line #385 for ostree image which does not support FS customizations at all, which would be IMO confusing.

Comment (say-paul) [reply pkg/distro/rhel9/imagetype.go:383]: Changed the ordering to return early incase fs customization not allowed

File: test/configs/filesystem-customizations.json
Comment (say-paul) [reply test/configs/filesystem-customizations.json]: added them

File: internal/pathpolicy/path_policy.go
Comment (say-paul) [reply internal/pathpolicy/path_policy.go]: This does not apply any more with the new approach

File: test/configs/ostree-filesystem-customizations.json
Comment (thozza) [inline test/configs/ostree-filesystem-customizations.json:56]: `/var/roothome` is missing...

**Review by thozza — COMMENTED** [review]
Thanks for the changes. Overall the PR looks good. I commented on two small things.

Before I approve, I would like to ask you to boot-test the `edge-raw-image` image with mountpoints listed in https://github.com/osbuild/images/pull/255#discussion_r1413887536.

@achilleas-k suggested that we could use the `edge-ami` image type to boot-test this in our CI. Which would hopefully require only allowing the FS customizations for the image type, listing it in the `test/config-map.json` for the specific config.

File: internal/pathpolicy/policies_test.go
Comment (say-paul) [reply internal/pathpolicy/policies_test.go:91]: Since most of them its mostly driven by `TestMountpointPolicies` I kept them out,
Adding them.

**Comment by say-paul** [comment]
> Thanks for the changes. Overall the PR looks good. I commented on two small things.
> 
> Before I approve, I would like to ask you to boot-test the `edge-raw-image` image with mountpoints listed in [#255 (comment)](https://github.com/osbuild/images/pull/255#discussion_r1413887536).
> 
> @achilleas-k suggested that we could use the `edge-ami` image type to boot-test this in our CI. Which would hopefully require only allowing the FS customizations for the image type, listing it in the `test/config-map.json` for the specific config.

I think we need to modify the rhel/imagtype.go to allow customization in edge-ami, do we really want to unlock that too?

File: test/configs/ostree-filesystem-customizations.json
Comment (say-paul) [reply test/configs/ostree-filesystem-customizations.json:60]: added now

**Comment by thozza** [comment]
> I think we need to modify the rhel/imagetype.go to allow customization in edge-ami, do we really want to unlock that too?

Yes, that's what I meant by "allowing the FS customizations for the image type". If your team does not feel comfortable with enabling the FS customizations for that image type, then please at least manually boot-test the image.

File: test/config-map.json
Comment (thozza) [inline test/config-map.json:197]: Typo: you are missing "s" in "customizations"

**Comment by say-paul** [comment]
> If your team does not feel comfortable with enabling the FS customizations for that image type, then please at least manually boot-test the image.
I am up for enabling it for edge-ami too, but I would defer to the teams decision on this which I will update shortly.

**Comment by say-paul** [comment]
>> If your team does not feel comfortable with enabling the FS customizations for that image type, then please at least manually boot-test the image.

>I am up for enabling it for edge-ami too , but I would defer to the teams decision on this which I will update shortly.

Looking into the broader epic that I am working,we need support for :
simplified-installer, edge-raw-image, edge-ami, edge-vsphere, so it makes sense to enable it for all .

**Comment by say-paul** [comment]
rebased and commit message updated

File: test/scripts/boot-image
Comment (achilleas-k) [inline test/scripts/boot-image:89]: This is fine but we might want to bios boot some images where supported (our AMIs actually support both).

File: test/config-map.json
Comment (achilleas-k) [inline test/config-map.json:208]: I'm fine with limiting these but I'd like to investigate what's wrong with 9.0 and 9.1.  If it's some weirdness that is distro/package version specific, it's probably fine, but if it's something we're doing, it might be a symptom of a bigger issue, so it would be good to know.

**Review by achilleas-k — APPROVED** [review]
LGTM.  A couple of comments for follow-up stuff.

**Comment by achilleas-k** [comment]
Created new issues from my comments for follow-up.

File: internal/pathpolicy/policies.go
Comment (cgwalters) [reply internal/pathpolicy/policies.go]: FWIW https://github.com/ostreedev/ostree/pull/3114 will change this wrt ostree systems (if enabled) - there's less of a need to allow-list specific mountpoints.

**Pull request merged**
